File: android/src/main/java/app/capgo/asset_cache/AssetCachePlugin.java
package app.capgo.asset_cache;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(name = "AssetCache")
public class AssetCachePlugin extends Plugin {

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private AssetCache implementation;

    @Override
    public void load() {
        implementation = new AssetCache(getContext().getApplicationContext());
    }

    @PluginMethod
    public void get(PluginCall call) {
        run(call, () ->
            implementation.get(
                call.getString("url"),
                call.getString("key"),
                call.getObject("headers", new JSObject()),
                call.getObject("revalidate", new JSObject())
            )
        );
    }

    @PluginMethod
    public void remove(PluginCall call) {
        run(call, () -> implementation.remove(call.getString("key"), call.getString("url")));
    }

    @PluginMethod
    public void clear(PluginCall call) {
        run(call, () -> implementation.clear());
    }

    @PluginMethod
    public void list(PluginCall call) {
        run(call, () -> implementation.list());
    }

    @PluginMethod
    public void getCacheSize(PluginCall call) {
        run(call, () -> implementation.getCacheSize());
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", implementation.getPluginVersion());
        call.resolve(ret);
    }

    @Override
    protected void handleOnDestroy() {
        executor.shutdownNow();
        super.handleOnDestroy();
    }

    private void run(PluginCall call, Action action) {
        executor.execute(() -> {
            try {
                call.resolve(action.run());
            } catch (Exception exception) {
                call.reject(exception.getMessage());
            }
        });
    }

    private interface Action {
        JSObject run() throws Exception;
    }
}

File: ios/Sources/AssetCachePlugin/AssetCachePlugin.swift
import Foundation
import Capacitor

@objc(AssetCachePlugin)
public class AssetCachePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AssetCachePlugin"
    public let jsName = "AssetCache"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "list", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCacheSize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = AssetCache()

    @objc func get(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("url is required")
            return
        }

        let key = call.getString("key")
        let headers = call.getObject("headers") as? [String: String] ?? [:]
        let revalidate = call.getObject("revalidate") ?? [:]

        Task {
            do {
                let result = try await implementation.get(url: url, key: key, headers: headers, revalidate: revalidate)
                resolve(call, result)
            } catch {
                reject(call, error)
            }
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        Task {
            do {
                let result = try implementation.remove(key: call.getString("key"), url: call.getString("url"))
                resolve(call, result)
            } catch {
                reject(call, error)
            }
        }
    }

    @objc func clear(_ call: CAPPluginCall) {
        Task {
            do {
                let result = try implementation.clear()
                resolve(call, result)
            } catch {
                reject(call, error)
            }
        }
    }

    @objc func list(_ call: CAPPluginCall) {
        Task {
            do {
                let result = try implementation.list()
                resolve(call, result)
            } catch {
                reject(call, error)
            }
        }
    }

    @objc func getCacheSize(_ call: CAPPluginCall) {
        Task {
            do {
                let result = try implementation.getCacheSize()
                resolve(call, result)
            } catch {
                reject(call, error)
            }
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }

    private func resolve(_ call: CAPPluginCall, _ data: [String: Any]) {
        DispatchQueue.main.async {
            call.resolve(data)
        }
    }

    private func reject(_ call: CAPPluginCall, _ error: Error) {
        DispatchQueue.main.async {
            call.reject(error.localizedDescription)
        }
    }
}
